src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1692 @@
+
+      thjm2  = thjm1;
+      thjm1  = thj;
+      bjm2   = bjm1;
+      bjm1   = bj;
+      zjm2   = zjm1;
+      zjm1   = zj;
+      dzjm2  = dzjm1;
+      dzjm1  = dzj;
+      d2zjm2 = d2zjm1;
+      d2zjm1 = d2zj;
+    }
+  }
+
+  /* Compute yerr (if step adaptivity enabled) */
+  if (!ark_mem->fixedstep)
+  {
+    retval = step_mem->fe(ark_mem->tcur + ark_mem->h, ark_mem->ycur,
+                          ark_mem->tempv2, ark_mem->user_data);
+    step_mem->nfe++;
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+
+    /* Estimate the local error and compute its weighted RMS norm */
+    cvals[0] = p8;
+    Xvecs[0] = ark_mem->yn;
+    cvals[1] = -p8;
+    Xvecs[1] = ark_mem->ycur;
+    cvals[2] = p4 * ark_mem->h;
+    Xvecs[2] = ark_mem->fn;
+    cvals[3] = p4 * ark_mem->h;
+    Xvecs[3] = ark_mem->tempv2;
+
+    retval = N_VLinearCombination(4, cvals, Xvecs, ark_mem->tempv1);
+    if (retval != 0) { return (ARK_VECTOROP_ERR); }
+
+    *dsmPtr = N_VWrmsNorm(ark_mem->tempv1, ark_mem->ewt);
+    lsrkStep_DomEigUpdateLogic(ark_mem, step_mem, *dsmPtr);
+  }
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_TakeStepRKL:
+
+  This routine serves the primary purpose of the LSRKStepRKL module:
+  it performs a single RKL step (with embedding, if possible).
+
+  The output variable dsmPtr should contain estimate of the
+  weighted local error if an embedding is present; otherwise it
+  should be 0.
+
+  The input/output variable nflagPtr is used to gauge convergence
+  of any algebraic solvers within the step.  As this routine
+  involves no algebraic solve, it is set to 0 (success).
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int lsrkStep_TakeStepRKL(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+{
+  int retval;
+  sunrealtype* cvals;
+  sunrealtype hmax, w1, bjm1, bjm2, mus, bj, ajm1, cjm1, temj, cj, mu, nu;
+  const sunrealtype p8 = SUN_RCONST(0.8), p4 = SUN_RCONST(0.4);
+  N_Vector* Xvecs;
+  ARKodeLSRKStepMem step_mem;
+
+  /* initialize algebraic solver convergence flag to success,
+     temporal error estimate to zero */
+  *nflagPtr = ARK_SUCCESS;
+  *dsmPtr   = ZERO;
+
+  /* access ARKodeLSRKStepMem structure */
+  retval = lsrkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  cvals = step_mem->cvals;
+  Xvecs = step_mem->Xvecs;
+
+  /* Compute dominant eigenvalue and update stats */
+  if ((step_mem->new_dom_eig))
+  {
+    retval = lsrkStep_ComputeNewDomEig(ark_mem, step_mem);
+    if (retval != ARK_SUCCESS) { return (retval); }
+  }
+
+  /* determine the number of required stages */
+  int ss = SUNRfloor(
```

Comment:
Actually, it is not. This line computes a strict lower bound for `ss`, and then gradually increases it with a for loop to ensure that the stability constrain is satisfied. The least value of ss for which the inequality holds is then assigned to `step_mem->req_stages`.

Reply:
As I stated in the first review, I don't think the loop here is needed. You're looking for the smallest $s$ for which $s^2 + s - 2 >= 2 * c$, where $c = h * sprad$. This quadratic equation has two roots, but only one is positive: $s = \frac{\sqrt{9 + 8 c} - 1}{2}$. I think this can be simplified to
```
int ss_root = SUNRceil((SUNRsqrt(SUN_RCONST(9.0) + SUN_RCONST(8.0) * SUNRabs(ark_mem->h) * step_mem->sprad) - ONE) / TWO);
if (ss_root > step_mem->stage_max_limit) { ... }
int ss = SUNMAX(2, ss_root);
```

Reply:
@Steven-Roberts, this is a very brilliant catch. Now, they are both evaluated with a similar fashion.